app/src/main/java/com/xmartlabs/template/controller/SessionController.java

```diff
@@ -27,59 +31,93 @@
+   * Then, the {@link Session} object is updated with the information from the Authentication response object
+   * @param authResponse the Authentication response from the service from which the session information will be updated
+   * @return the {@link Session} object just created or updated
+   */
+  @CheckResult
   @NonNull
   private Session getSessionFromAuthInformation(@NonNull AuthResponse authResponse) {
-    Session session = getSession();
-    if (session == null) {
-      session = new Session();
-    }
+    Session session = Optional.ofNullable(getSession())
+        .orElse(new Session());
     session.updateSession(authResponse);
     return session;
   }
 
+  /**
+   * Retrieves the current stored {@link Session}, if it exists
+   * Only upon first request the {@link Session} object will be queried from {@link SharedPreferences}.
+   * Then, it will be stored in memory for faster access
+   * @return the current {@link Session}, or {@code null} if none exists
+   */
+  @CheckResult
   @Nullable
   public Session getSession() {
-    String sessionJsonString = sharedPreferences
-        .getString(PREFERENCES_KEY_SESSION, null);
-    try {
-      return gson.fromJson(sessionJsonString, Session.class);
-    } catch (JsonParseException e) {
-      Timber.w(e, "Error while deserializing the stored session");
-      deleteSession();
-      return null;
-    }
+    return session.orElseGet(() -> {
+      String sessionJsonString = sharedPreferences.getString(PREFERENCES_KEY_SESSION, null);
+
+      return Exceptional.of(() -> gson.fromJson(sessionJsonString, Session.class))
+          .ifException(e -> {
+            Timber.w(e, "Error while deserializing the stored session");
+            deleteSession();
+          })
+          .get();
+    });
   }
 
+  /**
+   * Stores the {@code session} into the {@link SharedPreferences}
+   * @param session the {@link Session} object to be stored
+   * @return {@code Single<Session>} object. Upon subscription, it will only fail if the session could not be stored
+   */
   @CheckResult
-  public Completable setSession(@NonNull Session session) {
-    return Completable.create(subscriber -> {
+  @NonNull
+  public Single<Session> setSession(@NonNull Session session) {
+    return Single.fromCallable(() -> {
       String sessionJsonString = gson.toJson(session);
       boolean committed = sharedPreferences
-              .edit()
-              .putString(PREFERENCES_KEY_SESSION, sessionJsonString)
-              .commit();
+          .edit()
+          .putString(PREFERENCES_KEY_SESSION, sessionJsonString)
+          .commit();
+
       if (committed) {
-        subscriber.onComplete();
-      } else {
-        subscriber.onError(new RuntimeException("The session change could not be committed"));
+        this.session = Optional.of(session);
+        return session;
       }
+      throw new RuntimeException("The session change could not be committed");
```

Comment:
I don't like `RuntimeException`. Any idea of a more specific exception? @xmartlabs/android